app/main.py

```diff
@@ -86,129 +88,78 @@
-            red_knight["hp"] += red_knight["potion"]["effect"]["hp"]
-
-    # -------------------------------------------------------------------------------
-    # BATTLE:
-
-    # 1 Lancelot vs Mordred:
-    lancelot["hp"] -= mordred["power"] - lancelot["protection"]
-    mordred["hp"] -= lancelot["power"] - mordred["protection"]
-
-    # check if someone fell in battle
-    if lancelot["hp"] <= 0:
-        lancelot["hp"] = 0
-
-    if mordred["hp"] <= 0:
-        mordred["hp"] = 0
-
-    # 2 Arthur vs Red Knight:
-    arthur["hp"] -= red_knight["power"] - arthur["protection"]
-    red_knight["hp"] -= arthur["power"] - red_knight["protection"]
-
-    # check if someone fell in battle
-    if arthur["hp"] <= 0:
-        arthur["hp"] = 0
-
-    if red_knight["hp"] <= 0:
-        red_knight["hp"] = 0
+class Knight:
+    def __init__(self, knight: dict) -> None:
+        self.name = knight["name"]
+        self.power = knight["power"]
+        self.hp = knight["hp"]
+        self.protection = 0
+        items_in_backpack = ["armour", "weapon", "potion"]
+        self.backpack = {
+            key: knight[key] for key in items_in_backpack if key in knight
+        }
 
-    # Return battle results:
-    return {
-        lancelot["name"]: lancelot["hp"],
-        arthur["name"]: arthur["hp"],
-        mordred["name"]: mordred["hp"],
-        red_knight["name"]: red_knight["hp"],
+    def apply_armour(self) -> None:
+        if self.backpack["armour"]:
+            for armour in self.backpack["armour"]:
+                self.protection += armour["protection"]
+
+    def apply_weapon(self) -> None:
+        self.power += self.backpack["weapon"]["power"]
+
+    def apply_potion(self) -> None:
+        if self.backpack["potion"] is not None:
+            if "power" in self.backpack["potion"]["effect"]:
+                self.power += self.backpack["potion"]["effect"]["power"]
+
+            if "protection" in self.backpack["potion"]["effect"]:
+                self.protection += (
+                    self.backpack["potion"]["effect"]["protection"]
+                )
+
+            if "hp" in self.backpack["potion"]["effect"]:
+                self.hp += self.backpack["potion"]["effect"]["hp"]
+
+    def battle(self, other: Knight) -> None:
+        if isistance(other, Knight):
+            if other.power - self.protection > 0:
+                self.hp -= other.power - self.protection
+
+            if self.power - other.protection > 0:
+                other.hp -= self.power - other.protection
+
+            if self.hp <= 0:
+                self.hp = 0
+
+            if other.hp <= 0:
+                other.hp = 0
+
+
+def battle(base_knights_config: dict) -> dict:
+    # create a list with individual knights
+    knights = [value for value in base_knights_config.values()]
+    list_dict = []
+    for knight in knights:
+        list_dict.append(knight)
+    # create a list with knight objects
+    list_knights = []
```

Comment:
The variable name `list_knights` is misspelled as `list_kinghts` in the previous review. Please ensure consistency and clarity by using `list_knights`.